tests/network/mod.rs

```diff
@@ -439,28 +464,39 @@ where
         Q: IntoIterator<Item = TargetedMessage<D::Message, NodeUid>> + Debug,
     {
         for msg in msgs {
-            match msg.target {
-                Target::All => {
-                    for node in self.nodes.values_mut() {
-                        if node.id != sender_id {
-                            node.queue.push_back((sender_id, msg.message.clone()))
+            // Transform the queued message by means of the adversary.
+            let adv_msgs = self.adversary.push_message(MessageWithSender {
+                sender: sender_id,
+                tm: msg.clone(),
+            });
+            // Dispatch the result of the transformation.
+            for adv_msg in adv_msgs {
+                match adv_msg.tm.target {
+                    Target::All => {
+                        for node in self.nodes.values_mut() {
+                            if node.id != adv_msg.sender {
+                                node.queue
+                                    .push_back((adv_msg.sender, adv_msg.tm.message.clone()));
+                            }
                         }
+                        self.observer
+                            .queue
+                            .push_back((adv_msg.sender, adv_msg.tm.message));
                     }
-                    self.observer
-                        .queue
-                        .push_back((sender_id, msg.message.clone()));
-                    self.adversary.push_message(sender_id, msg);
-                }
-                Target::Node(to_id) => {
-                    if self.adv_nodes.contains_key(&to_id) {
-                        self.adversary.push_message(sender_id, msg);
-                    } else if let Some(node) = self.nodes.get_mut(&to_id) {
-                        node.queue.push_back((sender_id, msg.message));
-                    } else {
-                        warn!(
-                            "Unknown recipient {:?} for message: {:?}",
-                            to_id, msg.message
-                        );
+                    Target::Node(to_id) => {
+                        if self.adv_nodes.contains_key(&to_id) {
+                            self.adversary.push_message(MessageWithSender {
```

Comment:
Or should that even be a different method? This one is for messages sent to the adversary, while the other invocation is kind of eavesdropping on any communication.

Reply:
Or we could rename `push_message`. It is more of a `transform_message` now.